File: src/pack/Scenerio9.scala
package pack
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.sql._
import org.apache.spark.sql.types._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.expressions._
object Scenerio9 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("Scenerio9")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")
    val spark = SparkSession.builder().getOrCreate()
    import spark.implicits._

    val df = Seq(
      ("a", Seq(1, 1, 1, 3)),
      ("b", Seq(1, 2, 3, 4)),
      ("c", Seq(1, 1, 1, 1, 4)),
      ("d", Seq(3))).toDF("name", "rank")

    df.show()

    val explodedf = df.withColumn("rank", explode(col("rank")))
    explodedf.show()

    val filtdf = explodedf.filter(col("rank") === 1)
    filtdf.show()

    val countdf = filtdf.groupBy("name").agg(count("*").as("count")).orderBy(col("count") desc)
    countdf.show()

    val finaldf = countdf.select(col("name")).first().getString(0)
    println(finaldf)

  }
}

File: Scenerio9.py
from pyspark import SparkConf, SparkContext
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql.window import *

conf = SparkConf().setMaster("local[*]").setAppName("Scenerio-9")
sc = SparkContext(conf=conf)
sc.setLogLevel("ERROR")
spark = SparkSession.builder.getOrCreate()
data = [
    ("a", [1, 1, 1, 3]),
    ("b", [1, 2, 3, 4]),
    ("c", [1, 1, 1, 1, 4]),
    ("d", [3])
]
df = spark.createDataFrame(data, ["name", "rank"])
df.show()

explodedf = df.withColumn("rank", explode(col("rank")))
explodedf.show()

filtdf = explodedf.filter(col("rank") == 1)
filtdf.show()

countdf = filtdf.groupBy("name").agg(count("*").alias("count"))
countdf.show()

finaldf = countdf.select(col("name")).first()[0]
print(finaldf)
